File: apps/python-flask/src/routes/demo/dependencies.py
"""External dependency demonstration routes."""

from importlib import import_module
from time import perf_counter

import requests

from src.config.telemetry import get_logger


flask = import_module("flask")
Blueprint = flask.Blueprint
current_app = flask.current_app
g = flask.g
jsonify = flask.jsonify


logger = get_logger(__name__)
dependencies_bp = Blueprint("dependencies_demo", __name__, url_prefix="/api/dependencies")


@dependencies_bp.get("/external")
def external_dependency_demo():
    api_url = "https://jsonplaceholder.typicode.com/posts/1"
    logger.debug("Calling external API", extra={"custom_dimensions": {"url": api_url}})

    start = perf_counter()
    timeout = current_app.config["APP_SETTINGS"].external_api_timeout_seconds

    try:
        response = requests.get(api_url, timeout=timeout)
        duration_ms = round((perf_counter() - start) * 1000, 2)
        response.raise_for_status()
        data = response.json()

        logger.info(
            "External API call successful",
            extra={
                "custom_dimensions": {
                    "url": api_url,
                    "statusCode": response.status_code,
                    "duration": duration_ms,
                }
            },
        )

        return jsonify(
            {
                "message": "External dependency call successful",
                "data": data,
                "metadata": {
                    "duration": duration_ms,
                    "statusCode": response.status_code,
                    "correlationId": getattr(g, "correlation_id", None),
                },
            }
        )
    except requests.RequestException as exc:
        duration_ms = round((perf_counter() - start) * 1000, 2)
        logger.error(
            "External API call failed",
            exc_info=True,
            extra={
                "custom_dimensions": {
                    "url": api_url,
                    "error": str(exc),
                    "duration": duration_ms,
                }
            },
        )

        return (
            jsonify(
                {
                    "error": "Service Unavailable",
                    "message": "Failed to call external API",
                    "correlationId": getattr(g, "correlation_id", None),
                }
            ),
            503,
        )

File: apps/nodejs/src/routes/demo/dependencies.js
const express = require('express');
const router = express.Router();
const { logger } = require('../../config/telemetry');

router.get('/external', async (req, res) => {
  const apiUrl = 'https://jsonplaceholder.typicode.com/posts/1';

  logger.debug('Calling external API', {
    url: apiUrl,
    correlationId: req.correlationId,
  });

  const start = Date.now();

  try {
    const response = await fetch(apiUrl);
    const duration = Date.now() - start;
    const data = await response.json();

    logger.info('External API call successful', {
      url: apiUrl,
      statusCode: response.status,
      duration,
      correlationId: req.correlationId,
    });

    res.json({
      message: 'External dependency call successful',
      data,
      metadata: {
        duration,
        statusCode: response.status,
        correlationId: req.correlationId,
      },
    });
  } catch (error) {
    const duration = Date.now() - start;

    logger.error('External API call failed', {
      url: apiUrl,
      error: error.message,
      duration,
      correlationId: req.correlationId,
    });

    res.status(503).json({
      error: 'Service Unavailable',
      message: 'Failed to call external API',
      correlationId: req.correlationId,
    });
  }
});

router.get('/database-query', async (req, res) => {
  const queryName = req.query.query || 'getUserById';

  logger.debug('Simulated database query', {
    queryName,
    correlationId: req.correlationId,
  });

  const start = Date.now();
  await new Promise((resolve) => setTimeout(resolve, Math.random() * 200 + 50));
  const duration = Date.now() - start;

  // Simulate 10% failure rate
  if (Math.random() > 0.9) {
    logger.error('Database query failed', {
      queryName,
      error: 'Connection timeout',
      duration,
      correlationId: req.correlationId,
    });

    return res.status(500).json({
      error: 'Internal Server Error',
      message: 'Database query failed',
      correlationId: req.correlationId,
    });
  }

  logger.info('Database query successful', {
    queryName,
    duration,
    rowsReturned: 1,
    correlationId: req.correlationId,
  });

  res.json({
    message: 'Database query successful',
    queryName,
    metadata: {
      duration,
      rowsReturned: 1,
      correlationId: req.correlationId,
    },
  });
});

module.exports = router;
